lib/src/rules/invariant_booleans.dart

```diff
@@ -216,15 +216,35 @@ AstNodePredicate _isConditionalStatementWithReturn(
 AstNodePredicate _noFurtherAssignmentInvalidatingCondition(
     Expression node, Iterable<AstNode> nodesInDFS) {
   Set<Identifier> identifiers = _findStatementIdentifiers(node.parent);
-  return (AstNode statement) =>
-      nodesInDFS
-          .skipWhile((n) => n != statement)
-          .takeWhile((n) => n != node)
-          .where((n) =>
-              n is AssignmentExpression &&
-              !identifiers.contains(n.leftHandSide))
-          .length ==
-      0;
+  return (AstNode statement) {
+    bool isMutation(AstNode n) {
```

Comment:
This is fairly inefficient because you have to perform multiple `is` checks for every node, even when you've already determined that it isn't necessary. Perhaps re-structure slightly:
```
if (n is AssignmentExpression) {
  return !identifiers.contains(n.leftHandSide);
} else if (n is PostfixExpression) {
  TokenType type = n.operator.type;
  return (type == TokenType.PLUS_PLUS || type == TokenType.MINUS_MINUS) && !identifiers.contains(n.operand);
} else ...
}
```